File: plugin/claude-code/scripts/query.mjs
#!/usr/bin/env /usr/bin/node
/**
 * drymem query helper — called from Claude Code hook scripts.
 * Usage: query.mjs <project_path> <command>
 *   command: context | search <keyword>
 */
import { createRequire } from "module";
import { fileURLToPath } from "url";
import { dirname, resolve } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const require = createRequire(import.meta.url);

const DRYMEM_DIR =
  process.env.DRYMEM_DIR ?? resolve(__dirname, "../../..");
const DB_PATH =
  process.env.DRYMEM_DB_PATH ?? resolve(DRYMEM_DIR, "drymem.sqlite");

const projectPath = process.argv[2] ?? process.cwd();
const command = process.argv[3] ?? "context";
const keyword = process.argv[4] ?? "";

let db;
try {
  const Database = require(resolve(DRYMEM_DIR, "node_modules/better-sqlite3"));
  db = new Database(DB_PATH, { readonly: true });
} catch {
  process.exit(0);
}

try {
  if (command === "context") {
    const rows = db
      .prepare(
        `SELECT topic_key, content
         FROM memories
         WHERE project_path = ?
           AND deleted_at IS NULL
         ORDER BY updated_at DESC
         LIMIT 5`
      )
      .all(projectPath);

    if (rows.length > 0) {
      console.log(
        rows
          .map((r) => `### ${r.topic_key}\n${r.content}`)
          .join("\n\n---\n\n")
      );
    }
  } else if (command === "search" && keyword) {
    const rows = db
      .prepare(
        `SELECT topic_key, content
         FROM memories
         WHERE project_path = ?
           AND deleted_at IS NULL
           AND (topic_key LIKE '%' || ? || '%' OR content LIKE '%' || ? || '%')
         ORDER BY updated_at DESC
         LIMIT 5`
      )
      .all(projectPath, keyword, keyword);

    if (rows.length > 0) {
      console.log(
        rows
          .map((r) => `### ${r.topic_key}\n${r.content}`)
          .join("\n\n---\n\n")
      );
    }
  }
} catch {
  // Silent fail
} finally {
  db.close();
}

File: plugin/claude-code/scripts/query.py
#!/usr/bin/env python3
"""
drymem query helper for hook scripts.
Queries Neo4j directly (bypasses MCP) for fast context loading at session start.

Usage: query.py <project_path> context
"""

import hashlib
import os
import re
import sys
from datetime import datetime, timezone

from dotenv import load_dotenv

DRYMEM_DIR = os.environ.get("DRYMEM_DIR", os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))))
load_dotenv(os.path.join(DRYMEM_DIR, ".env"))


def sanitize_group_id(project_path: str) -> str:
    slug = re.sub(r"[^a-z0-9]", "-", project_path.lower().strip("/"))
    slug = re.sub(r"-+", "-", slug).strip("-")
    if len(slug) > 60:
        suffix = hashlib.sha1(project_path.encode()).hexdigest()[:8]
        slug = slug[:51] + "-" + suffix
    return slug


def get_context(project_path: str) -> str:
    try:
        from neo4j import GraphDatabase
    except ImportError:
        return ""

    uri = os.getenv("NEO4J_URI", "bolt://localhost:7687")
    user = os.getenv("NEO4J_USER", "neo4j")
    password = os.getenv("NEO4J_PASSWORD", "drymem_pass")
    group_id = sanitize_group_id(project_path)

    try:
        driver = GraphDatabase.driver(uri, auth=(user, password))
        with driver.session() as session:
            result = session.run(
                """
                MATCH (e:Episodic)
                WHERE e.group_id = $group_id
                RETURN e.name AS name, e.content AS content, e.created_at AS created_at
                ORDER BY e.created_at DESC
                LIMIT 5
                """,
                group_id=group_id,
            )
            records = list(result)
        driver.close()
    except Exception:
        return ""

    if not records:
        return ""

    lines = []
    for r in records:
        name = r["name"] or "untitled"
        content = (r["content"] or "")[:300]
        ts = r["created_at"] or ""
        lines.append(f"### {name} ({ts})\n{content}\n")

    return "\n".join(lines)


if __name__ == "__main__":
    if len(sys.argv) < 3:
        sys.exit(0)

    project_path = sys.argv[1]
    command = sys.argv[2]

    if command == "context":
        output = get_context(project_path)
        if output:
            print(output)
